### DISCUSSION TITLE
なぜDollarクラスのテストがDollarTestじゃなくてMoneyTestかついて

### DISCUSSION THREAD

**Discussion by rinahira** [code-polaris/reading-circle-tdd#50] [General]
TDD本を読んでいて、なぜDollarクラスのテストがDollarTestじゃなくてMoneyTestなんだろう？ってぼんやり疑問に思ってたのですが
[単体テストの考え方/使い方 ](https://www.amazon.co.jp/%E5%8D%98%E4%BD%93%E3%83%86%E3%82%B9%E3%83%88%E3%81%AE%E8%80%83%E3%81%88%E6%96%B9-%E4%BD%BF%E3%81%84%E6%96%B9-Vladimir-Khorikov/dp/4839981728) という本を読んでいたら答えが分かったのでシェアしますー。

- 単体(Unit)テストではUnitとしてどの単位で隔離するかの考え方で **古典学派**(デトロイト学派)と**ロンドン学派**(モック主義者)に分かれる

## 古典学派
- クラスではなく、一単位のふるまいをUnitとする
- クラス内で呼ばれる別クラス(協力者オブジェクト) もそのまま利用する
- 共有依存(DBなど他のテストケースと共有するもの) についてはモックする
- 古典学派の聖典は Kent Beckのテスト駆動開発 

## ロンドン学派
- クラスをUnitとする
- クラス内の協力者オブジェクトはモックする
- 不変依存(Enumとか変わらないもの)以外の依存すべてをモックする
- テストがテスト対象の詳細なコードに密結合してしまいがち
- ロンドン学派の聖典は [実践テスト駆動開発](https://www.amazon.co.jp/%E5%AE%9F%E8%B7%B5%E3%83%86%E3%82%B9%E3%83%88%E9%A7%86%E5%8B%95%E9%96%8B%E7%99%BA-Object-Oriented-SELECTION-Freeman/dp/4798124583)

Kent Beckのテスト駆動開発は古典学派なので、一単位のふるまいをテストケースにする
なのでDollarのテストはMoneyテスト(Dollarというクラスではなくお金のふるまいをテストしている)

きっとロンドン学派で同じテスト書こうとすると、
DollarクラスのテストはDollarTest, FrancクラスのテストはFrancTest みたいな感じになるのかなーと思いました。
自分は今まで意識せずロンドン学派寄りでテスト書いてたので、違いが気になってしまいました。本書読んで古典学派に宗旨替えしたい思ってます...

ちなみに この単体テストの考え方/使い方 の本、t-wadaさんが "**圧倒的名著**" とお勧めしてた本です。
(古典学派推しで、Kent Beckと同じ学派なので同時に読んでても今のところ混乱しません）
(結構ぶ厚くて全然読み途中ですが....)

**Comment by shino51** [comment]
え！！？それそんなに壮大な理由があったんですね！？
たまにツイッターなどで「モックはなるべくするな」っていう人がいるのが不思議だったのですが、それは古典学派の手腕を参考にしているということなのですかね。

古典学派のテストの仕方って、結合テストの考え方に近いような気がするのですが、古典学派の人達にとっての結合テストってどのくらいのスケールになるのかが少し気になります。

**Reply by rinahira** [reply]
>古典学派のテストの仕方って、結合テストの考え方に近いような気がするのですが、

確かに、ロンドン学派の考えだと複数クラスを検証する古典学派の単体テストは結合テストに見えますね～

>古典学派の人達にとっての結合テストってどのくらいのスケールになるのかが少し気になります。

古典学派的には
- 一単位の振る舞いを検証すること
- 実行時間が短いこと
- 他のテストケースから隔離された状態で実行すること

をすべて満たすのが単体試験で、一つでも満たさなかったら結合試験という扱いになるようです。
外部API接続やDB接続で時間がかかる試験とか、複数の振る舞いの連携を確認するような試験が該当するのではないかな～と思っております。

ECサイトのイメージだと
ロンドン学派の単体テスト → Cartクラス、Purchaseクラスなどをそれぞれテスト
古典学派の単体テスト → 購入という動作をテスト(購入数と在庫比較して購入可否判定、購入可なら在庫を減らすという一単位の振る舞い)
古典学派の結合テスト → カートに商品を入れる、購入、決済、購入後のメール送信 みたいな複数の振る舞いの連携テスト
みたいなイメージなのかな、と思いました！

**Comment by murnana** [comment]
学派の存在自体知らなかったです…！
勉強になります ✏️

**Comment by rinahira** [comment]
この辺の話、今後テスト駆動開発で触れられないのかなーとバーッと先読みしてみたら
巻末の付録C 訳者解説(t-wadaさんが書いた解説)で少し触れられていました。
気になる方は付録を先読みしてみてもよいかもです！

**Comment by kalen573** [comment]
今回のリーディングサークルで初めてTDDに触れていますが、
やりながら、自分は今何をテストしているんだろう？と思うことがしばしばありました。
こう言う事だったんですね。

「何をテストするか」に対するアプローチに色んな考え方や手法があり、
派閥まで分かれているとは驚きです。対立構造になってしまうと残念ですが、
こう言う対話や議論が業界を物凄いスピードで進化させているんだろうなと感じました。

こちらの話題を検索した時にヒットしたサイトを参考までに貼り付けておきます。

＜「古典派」TDD 対 「モック派」TDD、これは本当の対立軸だろうか？＞
https://www.infoq.com/jp/news/2009/03/classic-mockist-tdd/
